### ISSUE TITLE
Performance of the type checker is poor

### ISSUE THREAD

**Issue by dhil** [links-lang/links#359]
TL;DR Typing deeply nested recursive variants is very slow.

First a little bit of context: I wanted to implement a variation of PCF in Links to demonstrate various uses of effect handlers in compiler construction. I opted for a natural recursive variant type to represent the AST. However, I observed immense slow down in type checking the Links program, when I annotated small test programs with the AST-type. It took several minutes to type check.

The problem seems to be related to typing of recursive variants. It is easy to construct an artificial example that causes the type checker to take its sweet time
```
typename Term = [|Var:String|Lam:(String, Term)|];
sig nested_13 : () -> Term
fun nested_13() {
 Lam("1",
  Lam("2",
   Lam("3",
    Lam("4",
     Lam("5",
      Lam("6",
       Lam("7",
        Lam("8",
         Lam("9",
          Lam("10",
           Lam("11",
            Lam("12",
             Lam("13",
              Var("1"))))))))))))))
}
```
This program took nearly 3 minutes to type check on my workstation. By omitting the type signature, the program type checks instantaneously. The inferred type of the `nested_13` is
```
() -> [|Lam:(String, 
          [|Lam:(String, 
            [|Lam:(String, 
              [|Lam:(String, 
                [|Lam:(String, 
                  [|Lam:(String, 
                    [|Lam:(String, 
                      [|Lam:(String, 
                        [|Lam:(String, 
                          [|Lam:(String, 
                            [|Lam:(String, 
                              [|Lam:(String, 
                                [|Lam:(String, [|Var:String|_::Any|])|_::Any|])|_::Any|])|_::Any|])|_::Any|])|_::Any|])|_::Any|])|_::Any|])|_::Any|])|_::Any|])|_::Any|])|_::Any|])|_::Any|])|_::Any|]
```

I expect the slow down is caused by type equality testing, as the following program exhibits the same performance characteristics
```
fun nested_13() {
 Lam("1",
  Lam("2",
   Lam("3",
    Lam("4",
     Lam("5",
      Lam("6",
       Lam("7",
        Lam("8",
         Lam("9",
          Lam("10",
           Lam("11",
            Lam("12",
             Lam("13",
              Var("1"))))))))))))))
}

fun foo() { nested_13() == nested_13() }
```

As a little experiment, I tried type checking variations of the `nested` program, the graph below summarises my findings. The x-axis is the number of nested `Lam` constructors, and the y-axis is the execution time.
![typechecker](https://user-images.githubusercontent.com/1827113/42169909-fbc510fa-7e0c-11e8-809d-24dd4ca1d8e0.jpg)

Data
```
Depth, Time
1, 330
2, 322
3, 321
4, 321
5, 323
6, 330
7, 350
8, 443
9, 831
10, 2393
11, 8524
12, 35120
13, 148979
```
The execution time is almost constant for `x < 8`.

I expect implementation is either using an inefficient data structure or doing something silly, as typing a similar program in OCaml is instantaneous
```
type term = [ `Var of string | `Lam of string * term ]
let nested_13 : unit -> term 
   = fun () -> 
     `Lam ("1", 
       `Lam("2", 
          `Lam("3", 
             `Lam("4", 
                `Lam("5", 
                   `Lam("6", 
                      `Lam("7", 
                         `Lam("8", 
                            `Lam("9", 
                               `Lam("10", 
                                   `Lam("11", 
                                      `Lam("12", 
                                         `Lam("13", `Var("1"))))))))))))))
```

**Comment by jamescheney** [comment]
Which branch / commit is this observed in?  Specifically, pre- or post-merging #355, which could have affected the performance of typechecking/equivalence checking for recursive types either positively or negatively?  I had been thinking of proposing some simple performance regression tests for that reason.  So maybe this should be one.

**Comment by dhil** [comment]
> Which branch / commit is this observed in?

That was before #355. The results are similar after #355, the following is the data after a single run of each program 
```
Depth, Time
1, 384
2, 368
3, 377
4, 381
5, 382
6, 377
7, 406
8, 527
9, 930
10, 2637
11, 9655
12, 38086
13, 168348
```

**Comment by frank-emrich** [comment]
#355 didn't change anything within the type checker, it merely moved some of the IR traversals to their own file and supplied a visitor for types, which is currently not used (yet). Therefore, #355 should have zero influence on the performance of the type checker.

**Comment by jamescheney** [comment]
Ah, sorry, I forgot that #355 was not actually doing any de-cycling itself!  So never mind.  In any case it would be good to check this kind of behavior.

If type eqivalence/unification is the culprit then it could be that we are testing for bisimilarity of cyclic recursive types in a correct-but-exponential way, which paradoxically might become easier to control by adopting a non-cyclic representation.  But I don't know the details of how that is implemented so should probably keep quiet...

**Comment by frank-emrich** [comment]
Independently from this issue, I have suspended my work on de-cycling the recursive types for now. I understand most of the recursion-related code in unify.ml now, but I couldn't find the source for some bugs I couldn't get rid of. Thus, I now have a branch where I changed the usages of recursive types/rows accordingly, but their creation (i.e., what happens in unify.ml) is broken.

If we decide to change that code in unify.ml for performance reasons and end up with a  non-cyclic approach as the result, I'm happy to provide those changes to the rest of the codebases. But for now, I already spent more than a 1 1/2 weeks on this, it's not working as expected and the non-cyclic recursion by itself does not seem worth the effort for me to pursue this further.

**Comment by dhil** [comment]
Regardless, it is something we should prioritise to fix. Personally, I think it is mildly embarrassing that it takes Links several minutes to type check a program that implements a simple syntax-directed type checker for a simply-typed lambda calculus.

As a workaround, one can make the recursion shallow by using a set of auxiliary functions to construct a term, this seems to drastically lower the complexity of typing e.g.
```
sig lam : (String, Term) -> Term
fun lam(b, body) { Lam(b, body) }

sig var_ : (String) -> Term
fun var_(v) { Var(v) }

sig nested : () -> Term
fun nested() {
  lam("1", 
    lam("2",
       var("1")));
```

**Comment by dhil** [comment]
@frank-emrich and I just did some poor-man's performance debugging. It appears the problem is related to `unify_rec` in `unify.ml` as this function is called more than 21 million times during type checking of the `nested_13` function.

**Comment by dhil** [comment]
Type annotations at every level of the recursive term speeds up type checking, e.g.
```
fun nested_13() {
 Lam("1",
  Lam("2",
   ...,
     Var("1") : Term) : Term) .... : Term) : Term
}
```
Given this and my previous comment, I am fairly convinced that the problem has to do with unification of deep non-recursive variant types and recursive variant types.

**Comment by dhil** [comment]
Recursion may be a red-herring as without type annotations the expression `nested_13() == nested_13()` does not seem to invoke `unify_rec`, rather `unify_rows'` and `unify_presence'` are invoked an extraordinary number of times.

**Comment by slindley** [comment]
The issue is that the row unification algorithm unifies each pair of matching fields twice in order to account for missing fields on the left or right. This simplifies the implementation a little, but can lead to poor performance. The functions `extend_field_env` and `unify_field_envs` in `unify.ml` need refactoring slightly in order to first identify overlapping labels and then to perform unification just once when labels overlap.

**Comment by jamescheney** [comment]
Looks to me like the exponential behavior is just due to the two calls to `aux` in `unify_field_envs`, so it should suffice to split this into:

1. Traverse `lenv`, `renv` and handle all labels present in both.
2. Handle the labels present in `lenv` but not `renv`
3. Handle the labels present in `renv` but not `lenv`

This shouldn't require changing much code, just splitting the aux function into two cases (say, `handle_common_fields` and `handle_unmatched_fields`) and doing something like this:

```
handle_common_fields lenv renv;
handle_unmatched_fields lenv renv;
handle_unmatched_fields renv lenv
```

though probably it would be even better calculate the union of the domains of `lenv` and `renv` and visit each field just once rather than two or three times...

**Comment by dhil** [comment]
During the previous meeting, we discussed that a potential solution to this problem is to introduce a bit of nominal typing by giving aliases unique names internally such that we can compare names directly rather than unrolling.

**Comment by SimonJF** [comment]
This (along with #412) is becoming problematic for my work on the Elm framework (things are taking minutes to typecheck).

Is there any way I could help push the nominal typing and explicit recursive types work forward? I believe (@dhil) that you have some work on a patch for this already; could we perhaps try and merge separately to the Proper Modules work, or are the two too linked?

**Comment by dhil** [comment]
> could we perhaps try and merge separately to the Proper Modules work, or are the two too linked?

It can be done separately, in fact, that is how I intended to do it. I am happy to off-load it to you, if you want to implement it. I reckon it will take a couple of days to fully implement. I think we should fix #457 first to get rid of the exponential blow up in the size of mutually recursive types. We can discuss it in detail on Monday.

**Comment by dhil** [comment]
It seems the performance of the type checker has been improved (possibly by #528), at least for the cases described in this issue.
